src/App.js

```diff
@@ -23,49 +25,107 @@ const generateSquares = () => {
   }
 
   return squares;
-}
+};
 
 const App = () => {
-
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
+
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(null);
+
+  const onClickCallback = (id) => {
+    setSquares((squares) =>{
+      let newBoard = squares.map((square)=>{
+        for (let property of square){
+          if (property.id === id && property.value === ''){
+            if (currentPlayer === PLAYER_1){
+              property.value = PLAYER_1;
+            } else if (currentPlayer === PLAYER_2){
+              property.value = PLAYER_2;
+            }
+          }
+        }
+        return square;
+      });
+    setWinner(checkForWinner());
```

Comment:
Notice that checkForWinner uses the squares state, which won't reflect any updates from the setSquares call until the next render. Since this _is_ seeing the changes being made to the square data, we can infer that we are actually modifying both the previous state and the new state (cross-state contamination). This doesn't cause any major problems in this project, but it can cause problems in larger projects. We should make sure that we are only make changes in new data for state, and not making changes to previous state under React's nose.